### PR TITLE
image vulnz caching

### PR DESCRIPTION
**Pull request by aaron-prindle**

### CODE REVIEW

**Comment by aaron-prindle** [comment]
fixes #218

File: pkg/kritis/crd/securitypolicy/securitypolicy.go [hunk 1]
```diff
@@ -51,9 +52,26 @@ func ImageSecurityPolicies(namespace string) ([]v1beta1.ImageSecurityPolicy, err
 	return list.Items, nil
 }
 
+type Validator struct {
+	ImageToVulnz map[string][]metadata.Vulnerability
```
Comment (tejal29) [inline pkg/kritis/crd/securitypolicy/securitypolicy.go]: Add comment for Public types.
Lets make the cache private.

Comment (aaron-prindle) [reply pkg/kritis/crd/securitypolicy/securitypolicy.go]: done

File: Makefile [hunk 2]
```diff
@@ -123,11 +124,11 @@ clean:
 	rm -rf $(BUILD_DIR)
 .PHONY: integration
 integration: cross
-	go test -ldflags "$(GO_LDFLAGS)" -v -tags integration $(REPOPATH)/integration -timeout 5m -- --remote=true
+	go test -ldflags "$(GO_LDFLAGS)" -v -tags integration $(REPOPATH)/integration -timeout 5m -- --remote=true --gke-cluster-name=$(LOCAL_GKE_CLUSTER_NAME)
 
 .PHONY: integration-local
 integration-local:
-	go test -ldflags "$(GO_LDFLAGS)" -v -tags integration $(REPOPATH)/integration -timeout 5m -remote=false -gac-credentials=$(LOCAL_GAC_CREDENTIALS_PATH)
+	go test -ldflags "$(GO_LDFLAGS)" -v -tags integration $(REPOPATH)/integration -timeout 5m -remote=false -gac-credentials=$(LOCAL_GAC_CREDENTIALS_PATH) --gke-cluster-name=$(LOCAL_GKE_CLUSTER_NAME)
```
Comment (priyawadhwa) [inline Makefile]: since remote=false here, I don't think setting the gke-cluster-name will do anything

Comment (aaron-prindle) [reply Makefile]: yea forgot about the remote flag being used

File: pkg/kritis/crd/securitypolicy/securitypolicy.go [hunk 3]
```diff
@@ -68,9 +86,18 @@ func ValidateImageSecurityPolicy(isp v1beta1.ImageSecurityPolicy, image string,
...
+	var vulnz []metadata.Vulnerability
+	var err error
+	if _, ok := v.imageToVulnz[image]; ok {
+		glog.Infof("found cached vulnz for image %s", image)
+		vulnz = v.imageToVulnz[image]
+	} else {
+		glog.Infof("no cached vulnz for image %s, getting from api", image)
+		vulnz, err = client.GetVulnerabilities(image)
+		if err != nil {
+			return nil, err
```
Comment (tstromberg) [inline pkg/kritis/crd/securitypolicy/securitypolicy.go]: It was like this before, but now that the function is larger, do you mind annotating the error message here in some way? Such as:

"GetVulnerabilities(%s): %v", image, err
"failed to get vulnerabilities for %s: %v", image, err

I feel like ValidateImageSecurityPolicy does so much, that if a user ever gets an unhelpful error back from it, it's going to be difficult for us to analyze later what the source was. 

Context: https://golang.org/doc/effective_go.html#errors

"When feasible, error strings should identify their origin, such as by having a prefix naming the operation or package that generated the error. "

File: pkg/kritis/crd/securitypolicy/securitypolicy.go [hunk 4]
```diff
@@ -51,9 +52,26 @@ func ImageSecurityPolicies(namespace string) ([]v1beta1.ImageSecurityPolicy, err
 	return list.Items, nil
 }
 
+type Validator struct {
+	imageToVulnz map[string][]metadata.Vulnerability
```
Comment (tstromberg) [inline pkg/kritis/crd/securitypolicy/securitypolicy.go]: Since caching inevitably causes confusion for new developers, do you mind adding a comment that specifically calls out that this is used as a cache?

File: Makefile [hunk 5]
```diff
@@ -40,6 +40,7 @@ RESOLVE_TAGS_PACKAGE = $(REPOPATH)/$(RESOLVE_TAGS_PATH)
 RESOLVE_TAGS_KUBECTL_DIR = ~/.kube/plugins/resolve-tags
 
 LOCAL_GAC_CREDENTIALS_PATH ?= /tmp/gac.json
+LOCAL_GKE_CLUSTER_NAME ?= test-cluster-2
```
Comment (tstromberg) [inline Makefile]: This change seems unrelated to the pull description. Do you mind moving this into a second PR so that it can be reverted separately if needed?

File: pkg/kritis/crd/securitypolicy/securitypolicy.go (hunk 4 above)
Comment (aaron-prindle) [reply pkg/kritis/crd/securitypolicy/securitypolicy.go]: sure! done

File: pkg/kritis/crd/securitypolicy/securitypolicy.go (hunk 3 above)
Comment (aaron-prindle) [reply pkg/kritis/crd/securitypolicy/securitypolicy.go]: i've update this particular error with more info as suggested.  If tracing error messages becomes an issue, it make make sense to use [pkg/errors](https://github.com/pkg/errors) throughout kritis.

File: Makefile (hunk 5 above)
Comment (aaron-prindle) [reply Makefile]: done
